lib/Frontend/Frontend.cpp

```diff
@@ -288,9 +288,25 @@ void CompilerInstance::setupDiagnosticVerifierIfNeeded() {
   }
 }
 
+void CompilerInstance::setupDependencyTrackerIfNeeded() {
+  const auto &Invocation = getInvocation();
+  const auto &opts = Invocation.getFrontendOptions();
+
+  if (opts.InputsAndOutputs.hasDependencyTrackerPath() ||
+      !opts.IndexStorePath.empty() || opts.TrackSystemDeps) {
+    // Note that we're tracking dependencies even when we don't need to write
+    // them directly; in particular, -track-system-dependencies affects how
+    // module interfaces get loaded, and so we need to be consistently tracking
+    // system dependencies throughout the compiler.
+    createDependencyTracker(opts.TrackSystemDeps);
```

Comment:
It should be possible to internalize this API completely now. All the remaining callers will invoke `setup`. I think code completion might be a little tricky, but there's probably a quick fix there too.

Reply:
Ah, I didn't realise `createDependencyTracker` was being called elsewhere – let's expand the `FrontendOptions` to allow the dependency tracker to be enabled for non-system dependencies too. How does this look?